telemetry/telemetry.go

```diff
@@ -0,0 +1,406 @@
+}
+
+// countingSet implements a set that counts how many
+// times a key is inserted. It marshals to JSON in a
+// way such that keys are converted to values next
+// to their associated counts.
+type countingSet map[interface{}]int
+
+// MarshalJSON implements the json.Marshaler interface.
+// It converts the set to an array so that the values
+// are JSON object values instead of keys, since keys
+// are difficult to query in databases.
+func (s countingSet) MarshalJSON() ([]byte, error) {
+	type Item struct {
+		Value interface{} `json:"value"`
+		Count int         `json:"count"`
+	}
+	var list []Item
+
+	for k, v := range s {
+		list = append(list, Item{Value: k, Count: v})
+	}
+
+	return json.Marshal(list)
+}
+
+var (
+	// httpClient should be used for HTTP requests. It
+	// is configured with a timeout for reliability.
+	httpClient = http.Client{Timeout: 1 * time.Minute}
+
+	// buffer holds the data that we are building up to send.
+	buffer          = make(map[string]interface{})
+	bufferItemCount = 0
+	bufferMu        sync.RWMutex // protects both the buffer and its count
+
+	// updating is used to ensure only one
+	// update happens at a time.
+	updating bool
+	updateMu sync.Mutex
+
+	// updateTimer fires off the next update.
+	// If no update is scheduled, this is nil.
+	updateTimer   *time.Timer
+	updateTimerMu sync.Mutex
+
+	// disabledMetrics is a set of metric keys
+	// that should NOT be saved to the buffer
+	// or sent to the telemetry server. The value
+	// indicates whether the entry is temporary.
+	// If the value is true, it may be removed if
+	// the metric is re-enabled remotely later. If
+	// the value is false, it is permanent
+	// (presumably becaues the user explicitly
+	// disabled it) and can only be re-enabled
+	// with user consent.
+	disabledMetrics   = make(map[string]bool)
+	disabledMetricsMu sync.RWMutex
+
+	// instanceUUID is the ID of the current instance.
+	// This MUST be set to emit telemetry.
+	// This MUST NOT be openly exposed to clients, for privacy.
+	instanceUUID uuid.UUID
+
+	// enabled indicates whether the package has
+	// been initialized and can be actively used.
+	enabled bool
+
+	// maxBufferItems is the maximum number of items we'll allow
+	// in the buffer before we start dropping new ones, in a
+	// rough (simple) attempt to keep memory use under control.
+	maxBufferItems = 100000
+)
+
+const (
+	// endpoint is the base URL to remote telemetry server;
+	// the instance ID will be appended to it.
+	endpoint = "https://telemetry-staging.caddyserver.com/v1/update/"
+
+	// defaultUpdateInterval is how long to wait before emitting
+	// more telemetry data if all retires fail. This value is
+	// only used if the client receives a nonsensical value, or
+	// doesn't send one at all, or if a connection can't be made,
+	// likely indicating a problem with the server. Thus, this
+	// value should be a long duration to help alleviate extra
+	// load on the server.
+	defaultUpdateInterval = 1 * time.Hour
```

Comment:
±Add some random seconds to spread this.

Reply:
You still need a contingency in case your endpoint cannot be reached.

Like when 10k EC2 instances are started at once, being the cause and victim.

Reply:
At the _exact same moment_? Like, _within milliseconds_? That does seem unlikely. In fact, spreading them out might make them more likely to converge to about the same moment for sending updates.

Reply:
That's a *yes* to your question expressing disbelief. FYI, some gateways/routers do hold back and batch messages some milliseconds.

In any case, spreading this some seconds will do no harm, and as you said you avoid the convergence by adding variance once the client has reached the endpoint.

Reply:
Fair enough; I've revised this locally with:

```go
reply.NextUpdate = defaultUpdateInterval + time.Duration(rand.Intn(int(1*time.Minute)))
```

and will push soon.